http/src/main/java/io/micronaut/http/body/DefaultMessageBodyHandlerRegistry.java

```diff
@@ -139,26 +140,39 @@ private MediaTypeQualifier(Argument<?> type,
         @Override
         public <K extends BeanType<T>> Collection<K> filter(Class<T> beanType, Collection<K> candidates) {
             List<K> all = new ArrayList<>(candidates.size());
-            List<K> matchesMediaType = new ArrayList<>(candidates.size());
-            List<K> matchesAll = new ArrayList<>(candidates.size());
             for (K candidate : candidates) {
                 String[] applicableTypes = candidate.getAnnotationMetadata().stringValues(annotationType);
                 if (applicableTypes.length == 0) {
-                    matchesAll.add(candidate);
+                    all.add(candidate);
                     continue;
                 }
                 for (String mt : applicableTypes) {
                     if (mediaTypes.contains(new MediaType(mt))) {
-                        matchesMediaType.add(candidate);
+                        all.add(candidate);
+                        break;
                     }
                 }
             }
             // Handlers with a media type defined should have a priority
-            all.addAll(matchesMediaType);
-            all.addAll(matchesAll);
+            all.sort(Comparator.comparing(this::findOrder).reversed());
             return all;
         }
 
+        private int findOrder(BeanType<?> beanType) {
+            int order = 0;
+            String[] applicableTypes = beanType.getAnnotationMetadata().stringValues(annotationType);
+            int size = mediaTypes.size();
+            for (String mt : applicableTypes) {
+                int index = mediaTypes.indexOf(new MediaType(mt));
+                if (index == -1) {
+                    continue;
+                }
+                int compareValue = size - index; // First value should have the priority
```

Comment:
since `size` is the same for every handler, i think you can replace size with 0, if you initialize `order` with `MIN_VALUE`. You can also probably get rid of the `-`

Reply:
I had -1 one befor and somehow it doesn't work

Reply:
let's leave it this way